python/dgl/contrib/cugraph/nn/conv/relgraphconv.py

```diff
@@ -0,0 +1,386 @@
+        message_flow_graph_hg_csr_int32, message_flow_graph_hg_csr_int64)
+except ModuleNotFoundError:
+    from pylibcugraphops.aggregators.node_level import (
+        agg_hg_basis_post_bwd_int32, agg_hg_basis_post_bwd_int64,
+        agg_hg_basis_post_fwd_int32, agg_hg_basis_post_fwd_int64)
+    from pylibcugraphops.structure.graph_types import (
+        message_flow_graph_hg_csr_int32, message_flow_graph_hg_csr_int64)
+
+CUDA_SM_PER_BLOCK = 49152
+
+
+class RelGraphConvAgg(th.autograd.Function):
+    @staticmethod
+    def forward(ctx, g, fanout, num_rels, edge_types, feat, coeff):
+        """
+        Compute the forward pass of R-GCN aggregation layer.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph.
+        fanout : int
+            Maximum in-degree of nodes.
+        num_rels : int
+            Number of edge types in this graph.
+        edge_types : torch.Tensor
+            Tensor of the edge types.
+        feat : torch.Tensor, dtype=torch.float32, requires_grad=True
+            Input feature, shape: (num_src_nodes, in_feat).
+        coeff : torch.Tensor, dtype=torch.float32, requires_grad=True
+            Coefficient matrix in basis-decomposition for regularization,
+            shape: (num_rels, num_bases). It should be set to ``None``
+            when ``regularizer=None``.
+
+        Returns
+        -------
+        agg_output : torch.Tensor, dtype=torch.float32
+            Aggregation output, shape: (num_dst_nodes, num_rels * in_feat)
+            when ``regularizer=None``, and (num_dst_nodes, num_bases * in_feat)
+            when ``regularizer='basis'``.
+
+        """
+        if not g.device.type == "cuda":
+            raise TypeError(
+                f"dgl.contrib.cugraph.nn.RelGraphConv requires "
+                f"DGL graph on device cuda, but got {g.device}"
+            )
+        ctx.device = g.device
+        if not ctx.device == edge_types.device:
+            raise RuntimeError(
+                f"Expected graph and etypes tensor on the same device, "
+                f"but got {ctx.device} and {edge_types.device}"
+            )
+        if not ctx.device == feat.device:
+            raise RuntimeError(
+                f"Expected graph and feature tensor on the same device, "
+                f"but got {ctx.device} and {feat.device}"
+            )
+
+        if g.idtype == th.int32:
+            mfg_csr_func = message_flow_graph_hg_csr_int32
+            agg_fwd_func = agg_hg_basis_post_fwd_int32
+        elif g.idtype == th.int64:
+            mfg_csr_func = message_flow_graph_hg_csr_int64
+            agg_fwd_func = agg_hg_basis_post_fwd_int64
+        else:
+            raise TypeError(
+                f"Supported ID type: torch.int32 or torch.int64, "
+                f"but got {g.idtype}"
+            )
+        ctx.graph_idtype = g.idtype
+
+        _in_feat = feat.shape[-1]
+        indptr, indices, edge_ids = g.adj_sparse("csc")
```

Comment:
I have a quick question here.  
Does cugraph-ops have to rely on homogenous graph adj matrix `g.adj_sparse` to generate mfg?

Reply:
Yes, cugraph-ops does require a CSR representation of the graph. That is why we need the conversion here.

Reply:
So, does it mean we have to convert to a homogeneous graph first like here, https://github.com/dmlc/dgl/blob/0aa8310a4db072daba51187bbc6d1d725467ba35/examples/pytorch/rgcn/entity_sample.py#L99 to be able to adapt cugraph-ops aggregator? Could it possible to support multiple CSR representations for hetero-graph-RGCN?

Reply:
> You mean 1 CSR adjacency matrix per edge type, right?

Yes. 
> No, our aggregator interface does not support that.

I see. Thanks. Can it support COO format?

Reply:
No, not natively. Otherwise, we won't need the explicit conversion from COO to CSR in the wrapper.